include/crocoddyl/multibody/residuals/pair-collisions.hpp

```diff
@@ -0,0 +1,157 @@
+// All rights reserved.
+///////////////////////////////////////////////////////////////////////////////
+
+#ifndef CROCODDYL_MULTIBODY_RESIDUALS_PAIR_COLLISIONS_HPP_
+#define CROCODDYL_MULTIBODY_RESIDUALS_PAIR_COLLISIONS_HPP_
+
+#include "crocoddyl/core/residual-base.hpp"
+#include "crocoddyl/multibody/states/multibody.hpp"
+#include "crocoddyl/multibody/data/multibody.hpp"
+#include "crocoddyl/core/utils/exception.hpp"
+#include "pinocchio/multibody/geometry.hpp"
+#include "pinocchio/algorithm/jacobian.hpp"
+#include "pinocchio/algorithm/geometry.hpp"
+#include "pinocchio/multibody/fcl.hpp"
+
+namespace crocoddyl {
+
+/**
+ * @brief Pair collisions residual
+ *
+ * This residual function defines the euclidean distance between a geometric collision pair as 
+ * \f$\mathbf{r}=\mathbf{p}_1-\mathbf{p}_2^*\f$, where \f$\mathbf{p}_1,\mathbf{p}_2^*\in~\mathbb{R}^3\f$
+ * are the nearest points on each collision object with respect to the other object. One of the objects is
+ * a body frame of the robot, the other is an external object. Note that for the
+ * sake of fast computation, it is easier to define the collision objects as inflated capsules.
+ * Note also that the dimension of the residual vector is obtained from 3. Furthermore, the Jacobians 
+ * of the residual function are computed analytically.
+ *
+ * As described in `ResidualModelAbstractTpl()`, the residual value and its Jacobians are calculated by `calc` and
+ * `calcDiff`, respectively.
+ *
+ * \sa `ResidualModelAbstractTpl`, `calc()`, `calcDiff()`, `createData()`
+ */
+template <typename _Scalar>
+class ResidualModelPairCollisionsTpl : public ResidualModelAbstractTpl<_Scalar> {
+ public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  typedef _Scalar Scalar;
+  typedef MathBaseTpl<Scalar> MathBase;
+  typedef ResidualModelAbstractTpl<Scalar> Base;
+  typedef ResidualDataPairCollisionsTpl<Scalar> Data;
+  typedef ResidualDataAbstractTpl<Scalar> ResidualDataAbstract;
+  typedef StateMultibodyTpl<Scalar> StateMultibody;
+  typedef DataCollectorAbstractTpl<Scalar> DataCollectorAbstract;
+  typedef pinocchio::GeometryModel GeometryModel;
+  
+  typedef typename MathBase::Vector3s Vector3s;
+  typedef typename MathBase::VectorXs VectorXs;
+  typedef typename MathBase::MatrixXs MatrixXs;
+
+  /**
+   * @brief Initialize the pair collisions residual model
+   *
+   * @param[in] state  State of the multibody system
+   * @param[in] nu   Dimension of the control vector
+   * @param[in] geom_model   Pinocchio geometry model containing the collision pair
+   * @param[in] pair_id   Index of the collision pair in the geometry model
+   * @param[in] joint_id   Index of the nearest joint on which the collision link is attached
+   */
+  
+  ResidualModelPairCollisionsTpl(boost::shared_ptr<StateMultibody> state,
+                             const std::size_t& nu,
+                             boost::shared_ptr<GeometryModel> geom_model,
+                             const pinocchio::PairIndex& pair_id, 
```

Comment:
I think `PairIndex` is a primitives (i.e., integer), right. If so, then change the code as
```cpp
                              const pinocchio::PairIndex pair_id, 
```